src/itree.cpp

```diff
@@ -132,15 +160,32 @@ const char* permission_type_string(fs::perms p) {
     }
 }
 
+namespace {
+    template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+    template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>; // deduction guide, not needed in C++20
+}
+
+// Helper function to invoke the right match type (regex / glob) based on std::variant contents.
+
+auto flexible_match(const std::string& obj_name, const pattern_matcher& pattern_variant) -> bool
+{
+    auto matches = overloaded {
+      [](const auto &)->bool {return false;},
+      [&obj_name](const std::string& glob)->bool {return !fnmatch(glob.c_str(), obj_name.c_str(), 0);},
+      [&obj_name](const std::regex& rgx)->bool {return std::regex_match(obj_name, rgx);}
+    };
+    return std::visit(matches, pattern_variant);
+}
+
 // Convenience function to wrap the logic for matching/ignoring.
+// The following seems to match the logic for the command 'tree' most closely:
+
 auto object_matches(const fs::collection_entry& entry) -> bool {
-    if( exclude_matcher.has_value() && std::regex_match(entry.path().object_name().string(), exclude_matcher.value()) ) {
-        return false;
-    }
-    if( !std::regex_match(entry.path().object_name().string(), matcher) ){
-        return false;
-    }
-    return true;
+    const auto & obj_name = entry.path().object_name().string();
+
+    // apply --ignore* and --pattern* as logically AND'ed together
+    return ((is_defaulted(exclude_matcher) || !flexible_match(obj_name, exclude_matcher))
+          &&(is_defaulted(matcher) || flexible_match(obj_name, matcher)));
```

Comment:
Let's move the logical-and to the end of the first line so that the reader's eyes see it.